statsmodels/tsa/vecm/vecm.py

```diff
@@ -0,0 +1,2146 @@
+    dt = np.dot(du, temp)
+
+    # JP: the next part can be done much  easier
+    auind = np.argsort(au)
+    aind = np.flipud(auind)
+    a = au[aind]
+    d = dt[:, aind]
+
+    #  Compute the trace and max eigenvalue statistics
+    lr1 = np.zeros(neqs)
+    lr2 = np.zeros(neqs)
+    cvm = np.zeros((neqs, 3))
+    cvt = np.zeros((neqs, 3))
+    iota = np.ones(neqs)
+    t, junk = rkt.shape
+    for i in range(0, neqs):
+        tmp = trimr(np.log(iota-a), i, 0)
+        lr1[i] = -t * np.sum(tmp, 0)
+        lr2[i] = -t * np.log(1-a[i])
+        cvm[i, :] = c_sja(neqs - i, det_order)
+        cvt[i, :] = c_sjt(neqs - i, det_order)
+        aind[i] = i
+
+    result = Holder()
+    # estimation results, residuals
+    result.rkt = rkt
+    result.r0t = r0t
+    result.eig = a
+    result.evec = d
+    result.lr1 = lr1
+    result.lr2 = lr2
+    result.cvt = cvt
+    result.cvm = cvm
+    result.ind = aind
+    result.meth = 'johansen'
+
+    return result
+
+
+class VECM(tsbase.TimeSeriesModel):
+    """
+    Fit a VECM process
+    .. math:: \\Delta y_t = \\Pi y_{t-1} + \\Gamma_1 \\Delta y_{t-1} + \\ldots + \\Gamma_{k_ar-1} \\Delta y_{t-k_ar+1} + u_t
+    where
+    .. math:: \\Pi = \\alpha \\beta'
+    as described in chapter 7 of [1]_.
+
+    Parameters
+    ----------
+    endog_tot : array-like (nobs_tot x neqs)
+        2-d endogenous response variable.
+    exog: ndarray (nobs_tot x neqs) or None
+        Deterministic terms outside the cointegration relation.
+    exog_coint: ndarray (nobs_tot x neqs) or None
+        Deterministic terms inside the cointegration relation.
+    dates : array-like of datetime, optional
+        See :class:`statsmodels.tsa.base.tsa_model.TimeSeriesModel` for more
+        information.
+    freq : str, optional
+        See :class:`statsmodels.tsa.base.tsa_model.TimeSeriesModel` for more
+        information.
+    missing : str, optional
+        See :class:`statsmodels.base.model.Model` for more information.
+    diff_lags : int
+        Number of lags in the VEC representation
+    coint_rank : int
+        Cointegration rank, equals the rank of the matrix :math:`\\Pi` and the
+        number of columns of :math:`\\alpha` and :math:`\\beta`.
+    deterministic : str {"nc", "co", "ci", "lo", "li"}, default: "nc"
+        * "nc" - no deterministic terms
+        * "co" - constant outside the cointegration relation
+        * "ci" - constant within the cointegration relation
+        * "lo" - linear trend outside the cointegration relation
+        * "li" - linear trend within the cointegration relation
+
+        Combinations of these are possible (e.g. "cili" or "colo" for linear
+        trend with intercept)
+    seasons : int, default: 0
+        Number of seasons. 0 means no seasons.
+    first_season : int, default: 0
+        Season of the first observation.
+
+    References
+    ----------
+    .. [1] Lutkepohl, H. 2005. *New Introduction to Multiple Time Series Analysis*. Springer.
+    """
```

Comment:
This needs a `Notes` section with more explanation about the options.
e.g. the exog, deterministic and seasons and their relationship could be explained here.